Smoke Test

AbdulrhmnGhanem opened the pull request:
- The test runs in the CI to:
  - assert `boardsList` is rendered to the main page.
  - assert the `boardList` is not empty.
  - assert the number of boards on the page equal the number of boards pulled from running `yarn get-boards`.

- The result of this test is recorded on the [Cypress dashboard](https://dashboard.cypress.io/projects/pk2d9c/runs?branches=%5B%5D&committers=%5B%5D&page=1&status=%5B%5D&tags=%5B%5D&timeRange=%7B%22startDate%22%3A%221970-01-01%22%2C%22endDate%22%3A%222038-01-19%22%7D)

AbdulrhmnGhanem commented:
So it caught that there were no boards, the [video](https://dashboard.cypress.io/projects/pk2d9c/runs/2/specs/fe2fa908-0d4d-4082-8185-031b5aea1b1a/video) shows that.

kasbah commented:
Nice! Is it currently failing on purpose or is it showing that error we have been seeing?

AbdulrhmnGhanem commented:
Having a passing build and a failing one will make this easier; we can compare the logs and see what is going on. So we need a passing build I guess!

AbdulrhmnGhanem commented:
The failing behavior can't be reproduced locally as the smoke test is passing locally.

Also if you want to run the test locally run `npx cypress open` after running `yarn serve` to save the free tier builds to the CI. I'll add this to the readme.

kasbah commented:
Is it not reproducible when doing a production build either? What about when using the caching feature?

kasbah commented:
Also seeing this problem on master right now (but I turned auto-deploy of so it's not live on the main site)

- https://travis-ci.org/github/kitspace/kitspace/builds/725907029
- https://app.netlify.com/sites/kitspace/deploys/5f5a200f68744d174e3c07ee
- https://5f5a200f68744d174e3c07ee--kitspace.netlify.app/

AbdulrhmnGhanem commented:
> Is it not reproducible when doing a production build either? What about when using the caching feature?

When I run these commands to use production config
```bash
export PATH="$(pwd)/node_modules/bin:$PATH"
scripts/plug_versions
scripts/get_boards production cached
./configure production cached
ninja -j4
```
The build timeout instantaneously
```
Error: timed out to get version for https://github.com/kammce/SJTwo
    at Timeout._onTimeout (/home/prodigy/matrials/Hassle/Javascript/projects/kitspace/scripts/utils.js:7:14)
    at listOnTimeout (internal/timers.js:549:17)
    at processTimers (internal/timers.js:492:7)

```
Maybe something with my network. Did you try ruuning the test locally with caching?

AbdulrhmnGhanem commented:
> Yeah, that looks like a network problem. I did try out the caching when I first implemented it.

I mean running the new smoke test.

kasbah commented:
Just ran a cached production build on this branch. Got this:

```    
1) checks that boards are rendered to screen


  0 passing (5s)
  1 failing

  1) Smoke Test
       checks that boards are rendered to screen:

      AssertionError: Timed out retrying: Too many elements found. Found '104', expected '3'.
      + expected - actual

      -104
      +3
      
```

AbdulrhmnGhanem commented:
It's failing for a different reason [now](https://dashboard.cypress.io/projects/pk2d9c/runs/7/specs/df71cda6-aad0-47cc-885a-cbd45c114393/video); the number of boards in `boards.json` doesn't equal the number of boards rendered, this might be a hint for something!

AbdulrhmnGhanem commented:
Try running `npx cypress open` locally with the development configuration, it should assert the number of boards.

kasbah commented:
This is `npx cypress open`, still on the production:

![image](https://user-images.githubusercontent.com/206854/92804583-555a5b00-f3b0-11ea-99dc-1f1ab446d469.png)

This is the boards.json

```json
[
  {
    "id": "github.com/kitspace-forks/index/index_mobo",
    "summary": "Index pick and place machine motherboard",
    "fromMulti": "boards/github.com/kitspace-forks/index"
  },
  {
    "id": "github.com/kitspace-forks/index/index_ringLight",
    "summary": "Index pick and place machine ring light",
    "fromMulti": "boards/github.com/kitspace-forks/index"
  },
  {
    "id": "github.com/kitspace-forks/index/index_piggyback",
    "summary": "Index pick and place machine piggy back",
    "fromMulti": "boards/github.com/kitspace-forks/index"
  }
]
```
(I last built a preview for https://github.com/sphawes/index)

When I `rm -f build/.temp/boards.json` and re-run ninja boards.json is now an empty array.

AbdulrhmnGhanem commented:
> When I `rm -f build/.temp/boards.json` and re-run ninja boards.json is now an empty array.

Interesting!

AbdulrhmnGhanem commented:
The last commit will make errors more explicit.

kasbah commented on package.json:
```diff
@@ -34,12 +35,15 @@
     "watch": "./scripts/watch",
     "start": "./scripts/get_boards && npm run serve & npm run watch",
+    "smoke-test": "start-test serve 8080 cy:run",
+    "cy:run": "cypress run --record --key 0ef4cdd6-db15-4617-a080-eeb584c686f4",
```
Isn't that key supposed to be secret?

AbdulrhmnGhanem replied:
Yes. Though I don't intend to leave this instance of `Cypress` live after merging this branch. Also, it will require writing a script to parse the `ENV` var; I just took the easy route.

AbdulrhmnGhanem replied:
Now the `CYPRESS_KEY` in Travis settings holds the value of the current key.

Please, create a Cypress account and generate a new key: 
1. delete `projectId`from `cypress.json` 
2. run `npx cypress open` and choose the `Runs` tab and follow the steps there.

Then update the key in Travis settings.

kasbah replied:
Do I need to, or is it fine running on your key for now?

AbdulrhmnGhanem replied:
It's okay. Though keep in mind that the key is in the git log. Also, I think it's better to have one managing all resources.

AbdulrhmnGhanem commented on .travis.yml:
```diff
@@ -8,8 +8,6 @@ addons:
     packages:
       - inkscape
       - kicad
-env:
-  - CACHED_BUILD=cached
```
I think this is the source of the defect.

kasbah replied:
What do you mean? We should be able to turn the env var on and off as we wish, including in the .yml file.

AbdulrhmnGhanem replied:
If you look into the git history, you will find that I added it when I was trying to get the build to work with the cache without setting the environment variable directly on the Travis dashboard. We should be able to set the environment variables from the `.travis` file but honestly, I don't understand how Travis manages the build options.:sweat_smile:
Do you actively use the `.travis` file to set the environment variables? is it a must-have?

kasbah replied:
I normally set it on Travis settings but Ian Ross uses the .travis.yml to set that var. It shouldn't make any difference how it's being set. One just overrides the other. You can see it being overriden e.g. here: https://travis-ci.org/github/kitspace/kitspace/builds/728930222#L545

AbdulrhmnGhanem replied:
Here is a thing I didn't notice, even if this was the reason it shouldn't make the build fail when you run it locally. The fact that I can have the same build setup locally is annoying. Have you tried `git bisect`?

kasbah replied:
I am not quite sure what you are saying.  Why is it annoying that you _can_ have the same build setup locally? Did you mean _can't_?

It would make sense to be consistent I guess: to only pass the env var in at one point or grab it from the env everywhere. But you can set the env var locally too using `export` or `CACHED_BUILD=false ./configure production false` etc. 

I've used `git bisect` before but not on this branch or kitspace at all yet.

AbdulrhmnGhanem replied:
> I am not quite sure what you are saying. Why is it annoying that you _can_ have the same build setup locally? Did you mean _can't_?

Ah sorry for the typo!

> I've used `git bisect` before but not on this branch or kitspace at all yet.

I meant [a few days ago](https://github.com/kitspace/kitspace/pull/304#issuecomment-690748375) I asked you to run bisect of you have the time; to see when `boards.json` started not to have all the boards as you mentioned [here](https://github.com/kitspace/kitspace/pull/304#issuecomment-690733826).

The only hypothesis I could come up with is that the targets which need `boards.json` are already built and cached but in this case, there is no need to rebuild `boards.json` at all, not rebuild it as an empty array. Also, I haven't tried to test this hypothesis ye because not being able to configure a production configuration.
 
Also, I think it's actually good in this case to make the test fail and retrigger a fresh build without cache as it's invalid even though boards get rendered correctly to the homepage.

AbdulrhmnGhanem replied:
This seems like a false positive.[729546081](https://travis-ci.org/github/kitspace/kitspace/builds/729546081) is failing while [729546085](https://travis-ci.org/github/kitspace/kitspace/builds/729546085) is passing. Both are built from the same commit the only difference is that they are different caches.

kasbah replied:
Doesn't seem like a false positive to me. Seems like those two builds are showing exactly the problem we are trying to solve here.

AbdulrhmnGhanem commented on tasks/makeSearchIndex.js:
```diff
@@ -0,0 +1,40 @@
+if (require.main !== module) {
+  module.exports = function(config) {
+    const targets = ['build/.temp/search_index.json']
+    const boards = parseProjects(config, process.env.CACHED_BUILD)
```
I didn't change this from the previous state, though I skeptical wether the right env var is `CACHED_BUILD` or `cached_build`

kasbah replied:
What do you mean by previous state? It's definitely `CACHED_BUILD`.

AbdulrhmnGhanem replied:
Oh, I think the `CACHED_BUILD` is refering the `var` in `.travis.yml` and it's gone now so it should use `processArgs` instead.

kasbah replied:
We also set it and unset it on Travis settings: https://travis-ci.org/github/kitspace/kitspace/settings

AbdulrhmnGhanem replied:
It fails [now](https://travis-ci.org/github/kitspace/kitspace/builds/728945333#L612) for not enough arguments to the configure command!

kasbah replied:
Just use `process.env`? I don't understand why you changed that...

AbdulrhmnGhanem commented:
I meant removing these two lines is the false positive not the build failure

On Wed, Sep 23, 2020, 11:43 AM Kaspar Emanuel <notifications@github.com>
wrote:

> *@kasbah* commented on this pull request.
> ------------------------------
>
> In .travis.yml
> <https://github.com/kitspace/kitspace/pull/304#discussion_r493380461>:
>
> > -env:
> -  - CACHED_BUILD=cached
>
> Doesn't seem like a false positive to me. Seems like those two builds are
> showing exactly the problem we are trying to solve here.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kitspace/kitspace/pull/304#discussion_r493380461>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AI3OMSPLAZKAA2H6VFE267TSHG7KLANCNFSM4REHYARQ>
> .
>

AbdulrhmnGhanem commented:
I tried retriggering the build four times, before each one I deleted the cache from the `caches panel` for both #304 and `smoke-test`. Somehow it still finds a cached and rebuilds from it! The build steps were at most 1xx step! I only didn't build from the cache if ` CACHED_BUILD` was set to `false`.

To look thoroughly at what the build logs I triggered the build with several configurations and added the logs temporarily to git, you also can inspect them if you want. I will try to `grep` and `diff` them to see what is going on.

AbdulrhmnGhanem commented:
I cleaned the log files and added docs for the process of doing so.

The only thing I can see is steps happen out of sync for different builds. That is expected; ninja is configured to run in two jobs. Random build failures, due to concurrency, are awfully common. I'll try changing the build config to use a single job.

AbdulrhmnGhanem commented:
I don't know what [this](https://dashboard.cypress.io/projects/pk2d9c/runs/58/specs/d2c2766e-5aac-4f46-afc4-e6be54f0bbfe/video) even means! It happened after changing `./ninja -j 2` to `./ninja -j 1`.

AbdulrhmnGhanem commented:
After deleting the cache and retriggering build, it exited once again so this change(`-j 1`) didn't make a difference.

AbdulrhmnGhanem commented:
The build sometimes exits due to this [error](https://travis-ci.org/github/kitspace/kitspace/builds/731906756#L1210) and sometimes it never happens. I don't know why this could happen. Can you take a look at how the dependencies are specified in both `configure` and `makeSearchIndex.js`? Is there any problem with them?

kasbah commented:
I never noticed a problem with the way it was defined. Ill try and re-review it soon.

AbdulrhmnGhanem commented:
This pattern happens over and over again:
1. Delete the cache and set `CACHED_BUILD` to `false`
2. Trigger the build, the build fails due to the dependency error.
3. Re-trigger the build, the build doesn't fail, though the tests fail.
4. Re-trigger the build, the tests pass.
5. Set `CACHED_BUILD` to `cached` 
6. Re-trigger the build, the tests pass.
7. The tests pass for any subsequent builds.

kasbah commented:
I believe there were two things going on here:

1. An argument to `parseProjects` that should have been a boolean was a string, causing it to try to use a cache
2. There was an unpredictable dependency loop in our build when running from a cache:
    - `makeBoardsJson.js` is supposed to create boards.json, but it would just update it when cached
    - if it was cached it already existed and anything that depnded on it could read the old boards.json

I fixed these two issues. I think this is working now. The important part is really that it won't deploy if the empty boards error does happen again. 

@AbdulrhmnGhanem can you clean up this pull-request please? Rebase it, squash commits that do the same thing, remove all the log files etc?

AbdulrhmnGhanem commented:
I rebased on the [use-bom-for-search](https://github.com/kitspace/kitspace/compare/use-bom-for-search) branch. Though the tests have failed once again!

AbdulrhmnGhanem commented:
I have a question: do you only set the `CACHED_BUILD` to `true` after having a passing build, i.e., for fresh builds it isn't set to true?

kasbah commented:
Haven't been very consistent about that. I change it all the time. I set it to false when something needs updating that would normally be cached. Never set it to `true` but to `cached`.

AbdulrhmnGhanem commented:
Look at [this](https://travis-ci.org/github/kitspace/kitspace/builds/734389206), it's a fresh build, i.e., no Travis cache, and `CACHED_BUILD` is set to `false`. Both the build and the tests have [passed](https://travis-ci.org/github/kitspace/kitspace/builds/734389206#L6361), though it [failed](https://travis-ci.org/github/kitspace/kitspace/builds/734389206#L697) because of the `cat` command  can't find the `previous-boards.json` file. I'll add a check and see!

AbdulrhmnGhanem commented:
So Travis has a tight policy for command failure which is equivalent to setting `set -euxo pipefail` at the top of a bash script. I wasn't aware of that, sorry for the last few commits! :sweat_smile:

kasbah commented on registry.json:
```diff
@@ -1,6 +1,370 @@
 [
```
Shouldn't be changing this file.

kasbah commented on cypress/integration/smoke.spec.js:
```diff
@@ -0,0 +1,24 @@
+describe('Smoke Test', () => {
+  const boardsPath = './build/.temp/boards.json'
+
+  // This is a dummy commit to re-trigger the build!
```
Remove that line.

kasbah commented on cypress.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "baseUrl": "http://localhost:8080",
+  "projectId": "pk2d9c"
```
Shouldn't this be an env variable too?

AbdulrhmnGhanem replied:
It can be [either](https://docs.cypress.io/guides/dashboard/projects.html#Project-ID) way. Though the cypress dashboard is public so making this env var won't make a difference. Also, it's generated automatically by the cypress command to identify all the runs across different machines, if for some reason it changed I think it's good to break the setup for all users.

kasbah commented:
I think I'll have to change it when I switch to my own Cypress account? I'd rather change it via an env var.
